File: figures/src/python/classes/training_dynamics.py
#!/usr/bin/env python3
"""Class-based Python renderer for training-dynamics figures."""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Any

from matplotlib.figure import Figure

from figures.src.python.common import (
    REPO_ROOT,
    add_panel_labels,
    apply_publication_layout,
    configure_matplotlib,
    export_csv,
    mm_to_inches,
    project_resources,
    resolve_font_stack,
    source_data_mapping,
    validate_common_contract,
    write_manifest,
)


def _load_loss_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "model": row["model"],
                "display_order": int(row["display_order"]),
                "epoch": int(row["epoch"]),
                "split": row["split"],
                "loss": float(row["loss"]),
            }
            for row in reader
        ]


def _load_metric_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "model": row["model"],
                "display_order": int(row["display_order"]),
                "epoch": int(row["epoch"]),
                "auroc": float(row["auroc"]),
                "label_epoch": row["label_epoch"],
                "best_epoch": row["best_epoch"],
            }
            for row in reader
        ]


def build_source_data(
    spec: dict[str, Any], loss_rows: list[dict[str, Any]], metric_rows: list[dict[str, Any]]
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    ordered_loss = sorted(
        loss_rows,
        key=lambda item: (item["display_order"], item["split"] != "train", item["epoch"]),
    )
    ordered_metric = sorted(metric_rows, key=lambda item: (item["display_order"], item["epoch"]))
    mapping = source_data_mapping(spec)
    export_csv(
        REPO_ROOT / mapping["a"],
        ordered_loss,
        ["model", "display_order", "epoch", "split", "loss"],
    )
    export_csv(
        REPO_ROOT / mapping["b"],
        ordered_metric,
        ["model", "display_order", "epoch", "auroc", "label_epoch", "best_epoch"],
    )
    return ordered_loss, ordered_metric


def _style_map(theme: dict[str, Any], models: list[str]) -> dict[str, dict[str, Any]]:
    colors = [
        theme["palette"]["categorical"][0],
        theme["palette"]["categorical"][1],
        theme["palette"]["categorical"][2],
    ]
    markers = ["o", "s", "D"]
    return {
        model: {
            "color": colors[index % len(colors)],
            "marker": markers[index % len(markers)],
        }
        for index, model in enumerate(models)
    }


def _label_offsets() -> dict[str, tuple[float, float]]:
    return {
        "Foundation model": (0.35, 0.0),
        "Hybrid GNN": (0.35, -0.02),
        "CNN baseline": (0.35, -0.04),
    }


def _summary_text(metric_rows: list[dict[str, Any]], models: list[str]) -> str:
    lines = ["Best validation AUROC"]
    for model in models:
        best_row = max(
            (row for row in metric_rows if row["model"] == model),
            key=lambda item: (item["auroc"], -item["epoch"]),
        )
        lines.append(f"{model}: {best_row['auroc']:.2f}")
    return "\n".join(lines)


def create_figure(spec_path: Path) -> Figure:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    loss_rows = _load_loss_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    metric_rows = _load_metric_rows(REPO_ROOT / str(spec["data_inputs"][1]))
    panel_a_rows, panel_b_rows = build_source_data(spec, loss_rows, metric_rows)
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)

    import matplotlib.pyplot as plt

    fig, axes = plt.subplots(
        1,
        2,
        figsize=(
            mm_to_inches(float(spec["size"]["width_mm"])),
            mm_to_inches(float(spec["size"]["height_mm"])),
        ),
        constrained_layout=bool(theme["layout"]["constrained"]),
    )
    apply_publication_layout(fig, theme)

    model_names = [model for _, model in sorted({(r["display_order"], r["model"]) for r in panel_a_rows})]
    styles = _style_map(theme, model_names)

    panel_a = axes[0]
    for model in model_names:
        for split, linestyle, alpha, linewidth in (
            ("train", "--", 0.65, 1.4),
            ("validation", "-", 0.95, 1.8),
        ):
            subset = sorted(
                [row for row in panel_a_rows if row["model"] == model and row["split"] == split],
                key=lambda item: item["epoch"],
            )
            x_values = [row["epoch"] for row in subset]
            y_values = [row["loss"] for row in subset]
            panel_a.plot(
                x_values,
                y_values,
                color=styles[model]["color"],
                linestyle=linestyle,
                linewidth=linewidth,
                alpha=alpha,
            )
            if split == "validation":
                dx, dy = _label_offsets()[model]
                panel_a.text(
                    x_values[-1] + dx,
                    y_values[-1] + dy,
                    model,
                    fontsize=theme["typography"]["annotation_font_size_pt"],
                    color=theme["palette"]["neutral"][0],
                    ha="left",
                    va="center",
                )
    panel_a.text(
        1.02,
        0.95,
        "solid = validation",
        transform=panel_a.transAxes,
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
        ha="left",
        va="top",
    )
    panel_a.text(
        1.02,
        0.87,
        "dashed = train",
        transform=panel_a.transAxes,
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
        ha="left",
        va="top",
    )
    panel_a.set_title("Training and validation loss", loc="left", pad=2)
    panel_a.set_xlabel("Epoch")
    panel_a.set_ylabel("Loss")
    panel_a.set_xlim(1, max(row["epoch"] for row in panel_a_rows) + 1.1)
    panel_a.set_ylim(0.2, max(row["loss"] for row in panel_a_rows) + 0.12)
    panel_a.spines["top"].set_visible(False)
    panel_a.spines["right"].set_visible(False)

    panel_b = axes[1]
    for model in model_names:
        subset = sorted(
            [row for row in panel_b_rows if row["model"] == model],
            key=lambda item: item["epoch"],
        )
        x_values = [row["epoch"] for row in subset]
        y_values = [row["auroc"] for row in subset]
        panel_b.plot(
            x_values,
            y_values,
            color=styles[model]["color"],
            linewidth=1.8,
            marker=styles[model]["marker"],
            markersize=4.6,
        )
        final_row = next(row for row in subset if row["label_epoch"] == "yes")
        dx, dy = _label_offsets()[model]
        panel_b.text(
            final_row["epoch"] + dx,
            final_row["auroc"] + dy * 0.5,
            model,
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][0],
            ha="left",
            va="center",
        )
        best_row = next(row for row in subset if row["best_epoch"] == "yes")
        panel_b.scatter(
            best_row["epoch"],
            best_row["auroc"],
            s=70,
            color=styles[model]["color"],
            marker="*",
            edgecolor=theme["palette"]["neutral"][0],
            linewidth=0.5,
            zorder=4,
        )
    panel_b.text(
        0.98,
        0.04,
        _summary_text(panel_b_rows, model_names),
        transform=panel_b.transAxes,
        ha="right",
        va="bottom",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][0],
        bbox={
            "boxstyle": "round,pad=0.25",
            "facecolor": "white",
            "edgecolor": theme["palette"]["neutral"][1],
            "linewidth": 0.6,
        },
    )
    panel_b.set_title("Validation AUROC trajectory", loc="left", pad=2)
    panel_b.set_xlabel("Epoch")
    panel_b.set_ylabel("Validation AUROC")
    panel_b.set_xlim(1, max(row["epoch"] for row in panel_b_rows) + 1.1)
    panel_b.set_ylim(0.5, max(row["auroc"] for row in panel_b_rows) + 0.05)
    panel_b.spines["top"].set_visible(False)
    panel_b.spines["right"].set_visible(False)

    add_panel_labels(spec, theme, [panel_a, panel_b])
    return fig


def build_figure(spec_path: Path) -> dict[str, str]:
    spec, theme, font_policy, profile, _ = project_resources(spec_path)
    loss_rows = _load_loss_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    metric_rows = _load_metric_rows(REPO_ROOT / str(spec["data_inputs"][1]))
    resolved_font = resolve_font_stack(font_policy)
    fig = create_figure(spec_path)
    output_dir = REPO_ROOT / str(spec["renderers"]["python"]["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)
    stem = str(spec["figure_id"])
    svg_path = output_dir / f"{stem}.svg"
    pdf_path = output_dir / f"{stem}.pdf"
    png_path = output_dir / f"{stem}.png"
    fig.savefig(svg_path, format="svg")
    fig.savefig(pdf_path, format="pdf")
    fig.savefig(png_path, format="png", dpi=int(profile["preview_dpi"]))
    outputs = {
        "svg": str(svg_path.relative_to(REPO_ROOT)),
        "pdf": str(pdf_path.relative_to(REPO_ROOT)),
        "png": str(png_path.relative_to(REPO_ROOT)),
    }
    write_manifest(
        spec,
        profile,
        resolved_font,
        "python",
        spec_path,
        outputs,
        [
            "paired_loss_and_metric_panels",
            "split_encoding_not_color_only",
            "direct_labels_at_final_epoch",
            "best_checkpoint_markers",
            "vector_first_export",
        ],
        {
            "panel_count": len(spec.get("panels", [])),
            "direct_label_count": 6,
            "annotation_count": 9,
            "best_checkpoint_count": 3,
        },
    )
    return outputs


if __name__ == "__main__":
    raise SystemExit("Use figures/src/python/run_class_renderer.py")

File: figures/src/r/classes/training_dynamics.R
repo_root <- getOption("manuscript.figure.repo_root")
if (is.null(repo_root)) {
  stop("manuscript.figure.repo_root option is required")
}
source(file.path(repo_root, "figures", "src", "r", "common.R"))

load_loss_rows <- function(path) {
  rows <- read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  rows$model <- as.character(rows$model)
  rows$split <- as.character(rows$split)
  rows <- rows[order(rows$display_order, rows$split != "train", rows$epoch), ]
  rows
}

load_metric_rows <- function(path) {
  rows <- read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  rows$model <- as.character(rows$model)
  rows$label_epoch <- as.character(rows$label_epoch)
  rows$best_epoch <- as.character(rows$best_epoch)
  rows <- rows[order(rows$display_order, rows$epoch), ]
  rows
}

build_source_data <- function(spec, loss_rows, metric_rows) {
  mapping <- source_data_mapping(spec)
  write_source_csv(
    file.path(repo_root, mapping$a),
    loss_rows,
    c("model", "display_order", "epoch", "split", "loss")
  )
  write_source_csv(
    file.path(repo_root, mapping$b),
    metric_rows,
    c("model", "display_order", "epoch", "auroc", "label_epoch", "best_epoch")
  )
  list(loss = loss_rows, metric = metric_rows)
}

style_maps <- function(theme, models) {
  colors <- c(
    theme$palette$categorical[[1]],
    theme$palette$categorical[[2]],
    theme$palette$categorical[[3]]
  )
  shapes <- c(16, 15, 18)
  data.frame(
    model = models,
    color = colors[seq_along(models)],
    shape = shapes[seq_along(models)],
    stringsAsFactors = FALSE
  )
}

label_offsets <- function() {
  data.frame(
    model = c("Foundation model", "Hybrid GNN", "CNN baseline"),
    dx = c(1.18, 1.18, 1.18),
    dy = c(0.0, -0.02, -0.04),
    stringsAsFactors = FALSE
  )
}

summary_text <- function(metric_rows, models) {
  lines <- c("Best validation AUROC")
  for (model in models) {
    subset <- metric_rows[metric_rows$model == model, ]
    best <- subset[which.max(subset$auroc), , drop = FALSE][1, , drop = FALSE]
    lines <- c(lines, sprintf("%s: %.2f", model, best$auroc))
  }
  paste(lines, collapse = "\n")
}

create_plot <- function(spec_path) {
  spec <- load_yaml(spec_path)
  theme <- load_yaml(file.path(repo_root, "figures/config/project_theme.yml"))
  font_policy <- load_yaml(file.path(repo_root, "figures/config/font_policy.yml"))
  profiles <- load_yaml(file.path(repo_root, "figures/config/venue_profiles.yml"))
  registry <- load_yaml(file.path(repo_root, "figures/registry/classes.yml"))$classes
  profile <- profiles$profiles[[spec$target_profile]]
  class_entry <- registry[[spec$class_id]]
  validate_common_contract(spec, theme, font_policy, profile, class_entry)
  resolved_font <- resolve_font(font_policy)
  plot_theme <- build_theme(theme, resolved_font) +
    ggplot2::theme(
      legend.position = "none",
      plot.margin = ggplot2::margin(5.5, 22, 5.5, 5.5)
    )

  loss_rows <- load_loss_rows(file.path(repo_root, spec$data_inputs[[1]]))
  metric_rows <- load_metric_rows(file.path(repo_root, spec$data_inputs[[2]]))
  source_rows <- build_source_data(spec, loss_rows, metric_rows)

  models <- unique(metric_rows$model)
  styles <- style_maps(theme, models)
  offsets <- label_offsets()

  loss_plot_rows <- merge(source_rows$loss, styles, by = "model", sort = FALSE)
  loss_labels <- merge(
    loss_plot_rows[loss_plot_rows$split == "validation" & loss_plot_rows$epoch == max(loss_plot_rows$epoch), ],
    offsets,
    by = "model",
    sort = FALSE
  )
  loss_labels$label_x <- loss_labels$epoch + loss_labels$dx
  loss_labels$label_y <- loss_labels$loss + loss_labels$dy

  plot_a <- ggplot2::ggplot(loss_plot_rows, ggplot2::aes(x = epoch, y = loss, color = model, group = interaction(model, split))) +
    ggplot2::geom_line(ggplot2::aes(linetype = split, alpha = split), linewidth = 0.8, show.legend = FALSE) +
    ggplot2::geom_text(
      data = loss_labels,
      ggplot2::aes(x = label_x, y = label_y, label = model),
      inherit.aes = FALSE,
      hjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::annotate(
      "text",
      x = max(loss_plot_rows$epoch) + 1.18,
      y = max(loss_plot_rows$loss) - 0.04,
      label = "solid = validation",
      hjust = 1,
      vjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::annotate(
      "text",
      x = max(loss_plot_rows$epoch) + 1.18,
      y = max(loss_plot_rows$loss) - 0.12,
      label = "dashed = train",
      hjust = 1,
      vjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::scale_color_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_linetype_manual(values = c("train" = "dashed", "validation" = "solid")) +
    ggplot2::scale_alpha_manual(values = c("train" = 0.65, "validation" = 0.95)) +
    ggplot2::labs(title = "Training and validation loss", x = "Epoch", y = "Loss") +
    ggplot2::coord_cartesian(xlim = c(min(loss_plot_rows$epoch), max(loss_plot_rows$epoch) + 1.3), ylim = c(0.2, max(loss_plot_rows$loss) + 0.12), clip = "off") +
    plot_theme

  metric_plot_rows <- merge(source_rows$metric, styles, by = "model", sort = FALSE)
  metric_labels <- merge(
    metric_plot_rows[metric_plot_rows$label_epoch == "yes", ],
    offsets,
    by = "model",
    sort = FALSE
  )
  metric_labels$label_x <- metric_labels$epoch + metric_labels$dx
  metric_labels$label_y <- metric_labels$auroc + metric_labels$dy / 2
  best_rows <- metric_plot_rows[metric_plot_rows$best_epoch == "yes", ]

  plot_b <- ggplot2::ggplot(metric_plot_rows, ggplot2::aes(x = epoch, y = auroc, color = model, group = model)) +
    ggplot2::geom_line(linewidth = 0.8, show.legend = FALSE) +
    ggplot2::geom_point(ggplot2::aes(shape = model), size = 2.2, show.legend = FALSE) +
    ggplot2::geom_point(
      data = best_rows,
      ggplot2::aes(x = epoch, y = auroc, color = model),
      inherit.aes = FALSE,
      shape = 8,
      size = 3.2,
      show.legend = FALSE
    ) +
    ggplot2::geom_text(
      data = metric_labels,
      ggplot2::aes(x = label_x, y = label_y, label = model),
      inherit.aes = FALSE,
      hjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::annotate(
      "label",
      x = max(metric_plot_rows$epoch) + 0.9,
      y = min(metric_plot_rows$auroc) + 0.02,
      label = summary_text(metric_plot_rows, models),
      hjust = 1,
      vjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      label.size = 0.25,
      fill = "white",
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::scale_color_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_shape_manual(values = stats::setNames(styles$shape, styles$model)) +
    ggplot2::labs(title = "Validation AUROC trajectory", x = "Epoch", y = "Validation AUROC") +
    ggplot2::coord_cartesian(xlim = c(min(metric_plot_rows$epoch), max(metric_plot_rows$epoch) + 1.3), ylim = c(0.5, max(metric_plot_rows$auroc) + 0.05), clip = "off") +
    plot_theme

  figure_plot <- (plot_a + plot_b + patchwork::plot_layout(ncol = 2)) +
    patchwork::plot_annotation(tag_levels = "a") &
    ggplot2::theme(
      plot.tag = ggplot2::element_text(
        size = as.numeric(theme$panel_labels$font_size_pt),
        face = theme$panel_labels$font_weight,
        family = resolved_font$family
      )
    )
  list(
    plot = figure_plot,
    spec = spec,
    theme = theme,
    profile = profile,
    resolved_font = resolved_font
  )
}

build_figure <- function(spec_path) {
  payload <- create_plot(spec_path)
  spec <- payload$spec
  theme <- payload$theme
  profile <- payload$profile
  resolved_font <- payload$resolved_font
  figure_plot <- payload$plot
  output_dir <- file.path(repo_root, spec$renderers$r$output_dir)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  stem <- spec$figure_id
  svg_path <- file.path(output_dir, sprintf("%s.svg", stem))
  pdf_path <- file.path(output_dir, sprintf("%s.pdf", stem))
  png_path <- file.path(output_dir, sprintf("%s.png", stem))
  width_in <- mm_to_inches(as.numeric(spec$size$width_mm))
  height_in <- mm_to_inches(as.numeric(spec$size$height_mm))

  svglite::svglite(file = svg_path, width = width_in, height = height_in, bg = theme$export_defaults$background, system_fonts = list(sans = resolved_font$family))
  print(figure_plot)
  grDevices::dev.off()
  grDevices::cairo_pdf(file = pdf_path, width = width_in, height = height_in, bg = theme$export_defaults$background, family = resolved_font$family)
  print(figure_plot)
  grDevices::dev.off()
  ragg::agg_png(filename = png_path, width = width_in, height = height_in, units = "in", res = as.numeric(profile$preview_dpi), background = theme$export_defaults$background)
  print(figure_plot)
  grDevices::dev.off()

  outputs <- list(
    svg = sub(paste0("^", repo_root, "/"), "", svg_path),
    pdf = sub(paste0("^", repo_root, "/"), "", pdf_path),
    png = sub(paste0("^", repo_root, "/"), "", png_path)
  )
  write_manifest(
    repo_root,
    spec,
    profile,
    resolved_font,
    spec_path,
    "r",
    outputs,
    list(
      "paired_loss_and_metric_panels",
      "split_encoding_not_color_only",
      "direct_labels_at_final_epoch",
      "best_checkpoint_markers",
      "vector_first_export"
    ),
    list(
      panel_count = length(spec$panels),
      direct_label_count = 6,
      annotation_count = 9,
      best_checkpoint_count = 3
    )
  )
  outputs
}
